### PR TITLE
BUG: sparse: fix argmin/argmax when all entries are nonzero

### LINKED ISSUE #16929: BUG: `scipy.sparse.csc_matrix.argmin` returns wrong values
### Describe your issue.

`scipy.sparse.csc_matrix.argmin` seems to return wrong value when all elements are positive.

### Reproducing Code Example

```python
>>> import scipy
>>> x = scipy.sparse.coo_matrix(([1, 0.1, 1, 1], ([0, 0, 1, 1], [0, 1, 0, 1])), shape=(2, 2)).tocsc()
>>> x.todense()
matrix([[1. , 0.1],
        [1. , 1. ]])
>>> x.argmin()
2
>>> x.todense().argmin()
1
```

### Error message

```shell
No error messages.
```

### SciPy/NumPy/Python version information

1.9.1 1.23.0 sys.version_info(major=3, minor=9, micro=7, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by perimosocordiae**
#### Reference issue
Fixes gh-16929.

#### What does this implement/fix?
The special case for "all nonzeros" was incorrect, returning the flat index of the `data` array instead of the sparse matrix itself. See the linked issue for examples.

### CODE REVIEW

**Comment by perimosocordiae** [comment]
Test failure is unrelated: Test_SVDS_PROPACK.test_small_sigma2[float32]

**Comment by WarrenWeckesser** [comment]
It looks like this doesn't catch all the cases. The following is run with this pull request:
```
In [2]: import scipy

In [3]: scipy.__version__
Out[3]: '1.10.0.dev0+1308.ca45421'

In [4]: from scipy.sparse import csr_matrix

In [5]: w = csr_matrix([[9, 4], [3, 0]])

In [6]: w.A
Out[6]: 
array([[9, 4],
       [3, 0]])

In [7]: w.argmin()  # Should be 3
Out[7]: 1

In [8]: w.A.argmin()
Out[8]: 3
```

**Comment by perimosocordiae** [comment]
Good catch, looks like the `_find_missing_index` helper function expects the flattened indices to be in sorted order, which won't necessarily be true for all COO-format matrices.

**Comment by perimosocordiae** [comment]
Errors still appear to be unrelated.

**Comment by WarrenWeckesser** [comment]
I was running through some more examples, and ran into this:
```
In [71]: csr_matrix([[5, 5, 3], [4, 9, 10], [3, 4, 9]]).argmin()  # Should be 2
Out[71]: 6
```
The convention followed by `argmin` and `argmax` is that the index of the first occurrence of the extreme value is to be returned:
```
In [72]: csr_matrix([[5, 5, 3], [4, 9, 10], [3, 4, 9]]).A.argmin()  # Should be 2
Out[72]: 2
```

File: scipy/sparse/_coo.py [hunk 1]
```diff
@@ -533,7 +533,7 @@ def _sum_duplicates(self, row, col, data):
         # Assumes (data, row, col) not in canonical format.
         if len(data) == 0:
             return row, col, data
-        order = np.lexsort((row, col))
+        order = np.lexsort((col, row))
```
Comment (jjerphan) [inline scipy/sparse/_coo.py:536]:
```suggestion
        # First sorting wr.t. to rows, and then on columns.
        # This allows being consistent with NumPy which identifies
        # the argmin or argmax as the first element in the C-order
        # if there are several argmin or argmax candidates.
        order = np.lexsort((col, row))
```

File: scipy/sparse/_data.py [hunk 2]
```diff
@@ -261,37 +261,37 @@ def _arg_min_or_max(self, axis, out, op, compare):
...
+
+        size = num_row * num_col
+        if size == mat.nnz:
+            return int(mat.row[am]) * num_col + int(mat.col[am])
+
+        ind = mat.row * num_col + mat.col
+        ind.sort()
+        zero_ind = _find_missing_index(ind, size)
+        if m == zero:
+            return min(zero_ind, am)
```
Comment (jjerphan) [inline scipy/sparse/_data.py:293]:
```suggestion
        # At this stage, the max or min is >= 0.
        first_implicit_zero_index = _find_missing_index(ind, size)
        if m == zero:
            return min(first_implicit_zero_index, am)
```

File: scipy/sparse/_data.py [hunk 3]
```diff
@@ -261,37 +261,37 @@ def _arg_min_or_max(self, axis, out, op, compare):
...
+        if 0 in self.shape:
+            raise ValueError("Can't apply the operation to "
+                             "an empty matrix.")
+
+        if self.nnz == 0:
+            return 0
+
+        zero = self.dtype.type(0)
+        mat = self.tocoo()
+        mat.sum_duplicates()
```
Comment (jjerphan) [inline scipy/sparse/_data.py:276]: We might want to document the conversion to the canonical format properly.
```suggestion
        # Conversion to the canonical COO format.
        mat = self.tocoo()
        mat.sum_duplicates()
```

File: scipy/sparse/_data.py [hunk 4]
```diff
@@ -261,37 +261,37 @@ def _arg_min_or_max(self, axis, out, op, compare):
...
+        am = op(mat.data)
+        m = mat.data[am]
+        num_row, num_col = mat.shape
```
Comment (jjerphan) [inline scipy/sparse/_data.py:279]: Could we document `op` here to indicate that it is either argmax or argmin? Or could we just rename it to `argmin_or_argmax`?

Also, could we use more explicit names?
```suggestion
        argm_index = op(mat.data)
        argm_value = mat.data[argm_index]
        num_row, num_col = mat.shape
```

If this suggestion is accepted, the logic bellow needs to be adapted accordingly.

**Review by jjerphan — APPROVED** [review]
LGTM. I think we might want to document this implementation at a point or at least explain the canonical format somewhere (but this is orthogonal to the objectify of this PR IMO).

File: scipy/sparse/_data.py [hunk 5]
```diff
@@ -261,37 +261,37 @@ def _arg_min_or_max(self, axis, out, op, compare):
...
-                        zero_ind = _find_missing_index(ind, size)
-                        if m == zero:
-                            return min(zero_ind, am)
-                        else:
-                            return zero_ind
-
-        return self._arg_min_or_max_axis(axis, op, compare)
+        if 0 in self.shape:
+            raise ValueError("Can't apply the operation to "
+                             "an empty matrix.")
```
Comment (dschult) [inline scipy/sparse/_data.py:269]:
```suggestion
            raise ValueError("Can't apply the operation to an empty matrix.")
```

**Review by dschult — COMMENTED** [review]
Only a minor spacing comment. And I've looked at the code and it matches previous code except where it shouldn't.

**Comment by perimosocordiae** [comment]
I did a pass over the code to add comments and rename some variables to be more understandable. Thanks for taking the time to decipher the cryptic naming, @jjerphan !

**Comment by jjerphan** [comment]
You are welcome, @perimosocordiae! Thank you for writing comprehensible comments.

Failures on the CI are unrelated the changes made in this PR. I think we can merge this PR once it looks good to another maintainer.

**Review by tylerjereddy — APPROVED** [review]
Ok, I'm pretty happy with the number of reviews/approvals on this bug fix. I ran a bunch of the failing examples Warren posted in the original issue and all those that I tested now produce the expected values.

**Pull request merged**